x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/aggs/mapreduce/MapReduceAggregator.java

```diff
@@ -0,0 +1,118 @@
+
+    protected MapReduceAggregator(
+        String name,
+        AggregationContext context,
+        Aggregator parent,
+        Map<String, Object> metadata,
+        AbstractMapReducer<?, ?, ?> mapReducer,
+        List<ValuesSourceConfig> configs
+    ) throws IOException {
+        super(name, AggregatorFactories.EMPTY, context, parent, CardinalityUpperBound.NONE, metadata);
+
+        List<ValuesExtractor> extractors = configs.stream()
+            .map(c -> context.getValuesSourceRegistry().getAggregator(REGISTRY_KEY, c).build(c))
+            .collect(Collectors.toList());
+
+        this.extractors = extractors;
+        this.mapReducer = mapReducer;
+        this.bigArrays = context.bigArrays();
+        this.profiling = context.profiling();
+        this.mapReduceContextByBucketOrdinal = new LongObjectPagedHashMap<>(1, bigArrays);
+    }
+
+    @Override
+    public InternalAggregation buildEmptyAggregation() {
+        return new InternalMapReduceAggregation(name, metadata(), mapReducer, mapReducer.mapInit(bigArrays), profiling);
+    }
+
+    @Override
+    public final InternalAggregation[] buildAggregations(long[] owningBucketOrds) throws IOException {
+        InternalAggregation[] results = new InternalAggregation[owningBucketOrds.length];
+        for (int ordIdx = 0; ordIdx < owningBucketOrds.length; ordIdx++) {
+            results[ordIdx] = new InternalMapReduceAggregation(name, metadata(), mapReducer, getMapReduceContext(ordIdx), profiling);
+        }
+        return results;
+    }
+
+    @Override
+    protected LeafBucketCollector getLeafCollector(LeafReaderContext ctx, LeafBucketCollector sub) throws IOException {
+        return new LeafBucketCollectorBase(sub, null) {
+            @Override
+            public void collect(int doc, long owningBucketOrd) throws IOException {
+                mapReducer.map(extractors.stream().map(extractor -> {
+                    try {
+                        return extractor.collectValues(ctx, doc);
+                    } catch (IOException e) {
+                        throw new UncheckedIOException(e);
+                    }
+                }), getMapReduceContext(owningBucketOrd));
+            }
+        };
+    }
+
+    @Override
+    public void collectDebugInfo(BiConsumer<String, Object> add) {
+        add.accept("map_reducer", mapReducer.getWriteableName());
+    }
+
+    @Override
+    protected void doClose() {
+        // TODO: the initialized MapReduceContext objects could be releasables, if something goes wrong, who is cleaning them up??
+        Releasables.close(mapReduceContextByBucketOrdinal);
+    }
+
+    private Writeable getMapReduceContext(long bucketOrd) {
+        Writeable context = mapReduceContextByBucketOrdinal.get(bucketOrd);
+        if (context == null) {
+
+            // TODO: as long as we haven't clarified who is closing, we can't use big arrays from the context
+            context = mapReducer.mapInit(BigArrays.NON_RECYCLING_INSTANCE);
```

Comment:
You cannot do this. `NON_RECYCLING_INSTANCE` is bypassing our circuit breaker. This means you will never trip the circuit breaker for your `TransactionStore` objects. Even though they are using all the bigArray objects (which would integrate with the circuit breaker if using the actual context bigArray) the circuit breaker is effectively ignored. Meaning the node would OOM before it hits the breaker. 

I think the solution here is:
 - Use the actual breaker here
 - When serializing, change to a serializable object type that does not use the breaker
 - When reducing, initialize a "merging" object with the breaker that does all the merging
 - Close it when a new Internal agg is created (with the internal representation of the values in the internal agg being serializable).

Reply:
Again, I _want_ to use the request breakers. 

> I have been against the NON_RECYCLING_INSTANCE

I am against it as well. That's not the point!

> When serializing ... When reducing

As noted in the comment, it has to be ensure that this can never leak. As we can not be sure that 2 methods are really called. If I allocated a chunk in method A and I release it in B, I have to ensure both are really called in _every_ situation. Can we make this guarantee? I think we can _not_.

Have a look at `AggregationContext`: The method `addReleasable` registers _releasables_ to be closed, exception handling ensures that this is always the case. `AggregationReduceContext` does not have such counterpart. If something breaks in between, we can leak memory.

What you describe has been implemented, everything is in place to use the context breakers. However I took it back after reviewing the code. We can still use breakers within single calls, but not if it's used over several call until we ensured it memory is released even between calls. In the mapping phase this will hopefully be possible. In the reduce phase we can at least use it withing `reduceFinalize`. Before switching I will write tests to ensure all assumptions are valid.

Right now it is better to crash the process than to take the risk of introducing a leak. If a node crashes with an OOM it gets automatically restarted. That's better than a node that leaked memory and continues to run with limited resources.

Reply:
> Right now it is better to crash the process than to take the risk of introducing a leak. If a node crashes with an OOM it gets automatically restarted. That's better than a node that leaked memory and continues to run with limited resources.

I agree it's better to crash than to leak memory. But, this aggregation should not be merged until we figure out how to prevent both. Without respecting the circuit breaker, especially on an aggregation that uses so much memory, we are effectively adding an attack vector for bringing down clusters.

>  If a node crashes with an OOM it gets automatically restarted.

Is this strictly true in all environments? I would have thought this is only the case in a managed cloud environment?

> AggregationReduceContext does not have such counterpart. If something breaks in between, we can leak memory.

This is effectively why intermediate aggs objects that are created are usually mergeable representations of the final result. This may not be strictly possible given the size of the frequent item sets. 

How does the circuit breaker react with streamed objects? I am guessing it is cleared once all the bytes are read off the stream? This is a key issue as there is no way to pass in a recycled (and thus circuit-breaker aware) bigArrays object when instantiating an object off of the stream.

@nik9000 ^ do you know more here?


The reduce context does have bigArrays. So, the reduce method can instantiate a new `TransactionStore`, feed in the intermediary representation of the discovered sets on the shards, merge, and return without leaking. The key in this idea would be the intermediary objects could not use a recycled bigArray or they would be objects such that they don't need data structures dependent on bigarray at all.


I suppose it all comes down to that aggs has always assumed the intermediate results to be merged are significantly smaller than the originally aggregated documents. For this agg, this is not strictly the case correct @hendrikmuhs ?


_as an aside_
It would be nice to be able prune a bit before returning to the coordinating node to make sure that the data returned to the coordinating node is significantly smaller than the raw data. I know we have talked about this before and it may be a feature added later.

Reply:
> there is no way to pass in a recycled (and thus circuit-breaker aware) bigArrays

Recycling and circuit breaking are 2 different things, not 1. `NON_RECYCLING_INSTANCE` should be renamed to `NON_RECYCLING_AND_NON_CIRCUIT_BREAKING_INSTANCE`. 

I admit I assumed it is just not recycling, but indeed it is neither recycling nor breaking. However, it seems possible to create a non-recycling _but breaking_ instance. There are 2 parameters, the recycler and the breaker. I found 1 case - unfortunately a unit test - where the recycler is `null`, but the breaker got initialized.

> I suppose it all comes down to that aggs has always assumed the intermediate results to be merged are significantly smaller than the originally aggregated documents. 

Yes, that might be the explanation for `cardinality`, however the part where it still has the original instance in memory and does a deep-copy using `NON_RECYCLING_INSTANCE` looks dangerous to me.

Reply:
> 100%, if we could get access to the circuit breaker service (created in the node start), we could do this.

it can be retrieved from the big arrays instance we get from the context, note the comment:

```
public CircuitBreakerService breakerService() {   // TODO this feels like it is for tests but it has escaped
```

I found 1 usage (in production code) that gets the breakerService, it adjusts this breaker before allocating memory. We could do a similar thing and e.g. write a size before writing transactionstore and when we deserialize we adjust the breaker before reading. I am not sure about creating our own BigArray instance, because we need one for every `InternalAgg`, instead - until a proper solution is in place - we could keep the `NON_RECYCLING_INSTANCE` but surround its usage with breaker adjustments.

Am I right that a _new_ breaker gets instantiated for every request? With other words this should be possible without bad side effects on other functionality?

Reply:
> Am I right that a new breaker gets instantiated for every request? With other words this should be possible without bad side effects on other functionality?

breakerService gives you access to any named breaker. These are singletons running on the node. They are the same everywhere and threadsafe.

I think on the "..Aggregator" side, using bigArrays shouldn't be that big of an issue. The issue regarding leaks and large data occurs after things are serialized back to the coordinator.